lib/llm/src/http/service/metrics.rs

```diff
@@ -120,13 +143,95 @@ impl Metrics {
         )
         .unwrap();
 
+        let buckets_ttfb = vec![
+            0.0, 0.025, 0.05, 0.1, 0.15, 0.2, 0.3, 0.4, 0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0,
+            3.0, 4.0, 8.0, 16.0, 32.0, 64.0, 128.0, 256.0,
+        ];
+        let ttfb = HistogramVec::new(
+            HistogramOpts::new(
+                format!("{}_http_service_ttfb_seconds", prefix),
+                "Time to first byte for LLM requests",
+            )
+            .buckets(buckets_ttfb.clone()),
+            &["model"],
+        )
+        .unwrap();
+
+        let mut ibl = Vec::new();
+        for p in [0.5, 0.9, 0.99] {
+            let ibl_p = HistogramVec::new(
+                HistogramOpts::new(
+                    format!(
+                        "{}_http_service_ibl_p{}_seconds",
+                        prefix,
+                        p.to_string().replace('.', "_")
+                    ),
+                    format!(
+                        "Time to inter byte latency for LLM requests at {} percentile",
+                        p
+                    ),
+                )
+                .buckets(buckets_ttfb.clone()),
+                &["model"],
+            )
+            .unwrap();
+            ibl.push(ibl_p);
+        }
+
         Metrics {
             request_counter,
             inflight_gauge,
             request_duration,
+            ttfb,
+            ibl_p50: ibl[0].clone(),
+            ibl_p90: ibl[1].clone(),
+            ibl_p99: ibl[2].clone(),
+            rolling_ttfb: Mutex::new(HashMap::new()),
         }
     }
 
+    fn record_ttfb(&self, model: &str, duration: Duration) {
+        let mut per_model_ttfb = self.rolling_ttfb.lock().unwrap();
+        let model_completions = per_model_ttfb
+            .entry(model.to_string())
+            .or_insert_with(VecDeque::new);
+
+        let now = Instant::now();
+        model_completions.push_back((now, duration));
+    }
+
+    /// get recent ttfb times for the given model
+    pub fn get_recent_ttfb_times(&self, model: &str) -> (Option<Duration>, Option<Duration>) {
+        let mut per_model_ttfb = self.rolling_ttfb.lock().unwrap();
+        if let Some(times) = per_model_ttfb
+            .get_mut(model)
+            .filter(|times| !times.is_empty())
+        {
+            // Prune to last 60 seconds
+            let now = Instant::now();
+            while let Some((start_time, _)) = times.front() {
+                if now.duration_since(*start_time) > Duration::from_secs(60) {
+                    times.pop_front();
+                } else {
+                    break;
+                }
+            }
```

Comment:
I think this can be simplified as:
```suggestion
            times.retain(|(start_time, _)| now.duration_since(*start_time) <= Duration::from_secs(60));
```

Reply:
worked